Dan-Mark-SceneManager

MV1090 opened the pull request:
We're Almost finished, All buttons have functionality and all work correctly with exception of the play again button, for this we need some sort of ResetGame() function in the GameManager. The score changes depending on the GameManager score, the Lives images change depending on the GameManager Lives, all seems to work as expected.

I noticed some errors in the script, and went back in to fix it. Hopefully this branch is all good and ready to merge.

Added a restart game method and high score method to the Game manger and added their functionality into the scene manager.

Commit:
Dan-Mark-SceneManager

We're Almost finished, All buttons have functionality and all work correctly with exception of the play again buutton, for this we need some sort of ResetGame() function in the GameManager. The score changes depending on the GameManager score, the Lives images change depending on the GameManager Lives, all seems to work as expeccted.

Commit:
Dan-Mark-SceneManager

I noticced some erros in script that stopped some of the buttons working correctly, Adjustments have been made now everything works how it should again.

Commit:
Estelle-Dan-Mark-changes made to SceneManager and GameManager

Added a restart game method and high score method to the Game manger and added their functionality into the scene manager.

EstelleSchoeman reviewed (approved):
I checked the code and tested the buttons on Unity, all seems good.

SolomanDavis commented on Assets/_Scripts/Managers/CanvasManager.cs:
```diff
@@ -0,0 +1,186 @@
+
+    [Header("Text")]
+    public TMP_Text scoreText;
+    public TMP_Text HighScoreText;
```
nit - public fields should all be Capitalized camel case. In either case it'd be best to make field names consistent (so either make all field names Capitalized camel case or change this line to `highScoreText`.

SolomanDavis commented on Assets/_Scripts/Managers/CanvasManager.cs:
```diff
@@ -0,0 +1,186 @@
+
+              default:
+              break;
+        }
```
While this works, it is best to avoid code repetition and do this in a more readable and cleaner way using arrays.
Here's an idea of the suggestion:
```csharp
    // This will be exposed as a list of Images in the Unity editor.
    // Each image should be added as an element (in order: 1Life, 2Life, 3Life) to the list in Unity.
    [Header("Image")]
    public Image[] LifeImages;

   // ...

    // Update is called once per frame
    void Update()
    {
        // ...

        // Sets appropriate life images to display according to number of player lives remaining.
        for (int i = 0; i < LifeImages.Length; ++i)
        {
            LifeImages[i].enabled = i < GameManager.Instance.Lives;
        }

        // ...
    }
```

Rather than a long switch statement, this just goes through through each `LifeImage` and enables it or disables it depending on the number of lives that are left.

Even better, this doesn't need to be done on every frame update. It only needs to be done when the number of lives changes meaning when `GameManager.Instance.Lives` changes. You could hook this up by writing a function here `UpdateLifeHUD(lives)` that is called by the `GameManager` whenever the number of lives changes.
```csharp
public void UpdateLifeHUD(int lives)
{
    for (int i = 0; i < LifeImages.Length; ++i)
    {
        LifeImages[i].enabled = i < lives;
    }
}
```

SolomanDavis commented on Assets/_Scripts/Managers/GameManager.cs:
```diff
@@ -1,14 +1,18 @@
     [SerializeField] public GameObject PlayerPrefab;
 
-    private int _score;
+    private int _HighScore;
```
nit - private fields should be lower camel case: `_highScore`

SolomanDavis commented on Assets/_Scripts/Managers/GameManager.cs:
```diff
@@ -1,14 +1,18 @@
-    private int _score;
+    private int _HighScore;
+
+    public int _score;
```
this field should be private

SolomanDavis commented on Assets/_Scripts/Managers/GameManager.cs:
```diff
@@ -69,10 +84,48 @@ protected override void Start()
+    // When called in canvas manager it changes the scene. 
+    public void ChangeScene(int buildIndex)
+    {
+        SceneManager.LoadScene(buildIndex);
```
I think using the explicit name of the scene is a lot clearer and easier to reason about and also means if the build settings change and the order in which the scenes are loaded is changed, this method isn't affected.

[Can use the overload for `LoadScene`](https://docs.unity3d.com/ScriptReference/SceneManagement.SceneManager.LoadScene.html):
```csharp
SceneManager.LoadScene(sceneName);
```

SolomanDavis reviewed (changes requested):
Functionality looks good but I had a few comments around repetition and others (nit just means something nitpicky, suggesting a formatting or naming change but I'm not going to block the PR so up to you)

My main suggestion was around the way the life images are updated. While it works, it can be done in far less lines of code if we use an array, check out the comment for an example of what I mean

Commit:
Dan-Mark-Ziyad-Estelle changes made

Changes made based around comments, everything should be formatted corrrectly and functionality works.
